File: tmp_saga_analyze.mjs
import { spawnSync } from "child_process";
import { parseCandles, technicalAnalysis, fundingBias, orderBookImbalance, tickerChangePct } from "./dist/blofin/analytics.js";

const py = "C:\\Users\\mknig\\AppData\\Local\\Programs\\Python\\Python312\\python.exe";
const cwd = "C:\\Users\\mknig\\owl-swarm";

function bridgeFile(method, file) {
  const r = spawnSync(py, ["blofin_bridge.py", method, `@${file}`], { cwd, encoding: "utf8" });
  const s = r.stdout.trim();
  const obj = s.lastIndexOf("{");
  const arr = s.indexOf("[");
  const i = obj >= 0 && (arr < 0 || obj < arr) ? obj : arr;
  return JSON.parse(s.slice(i));
}

const funding = bridgeFile("get_funding_rate", `${cwd}\\tmp_saga_funding.json`);
const book = bridgeFile("get_order_book", `${cwd}\\tmp_saga_book.json`);
const c1m = bridgeFile("get_candles", `${cwd}\\tmp_saga_1m.json`);
const c15 = bridgeFile("get_candles", `${cwd}\\tmp_saga_15m.json`);
const c1h = bridgeFile("get_candles", `${cwd}\\tmp_saga_1h.json`);
const tickerRaw = bridgeFile("get_ticker", `${cwd}\\tmp_saga_ticker.json`);
const ticker = Array.isArray(tickerRaw) ? tickerRaw[0] : tickerRaw;

const fr = Number(funding.fundingRate);
const fb = fundingBias(fr);
const ob = orderBookImbalance(book);
const bid = Number(book.bids[0][0]);
const ask = Number(book.asks[0][0]);
const spreadPct = ((ask - bid) / ((ask + bid) / 2)) * 100;

const candles1m = parseCandles(c1m);
const candles15 = parseCandles(c15);
const candles1h = parseCandles(c1h);
const ta1m = technicalAnalysis(candles1m, fr);
const ta15 = technicalAnalysis(candles15, fr);
const ta1h = technicalAnalysis(candles1h, fr);

const price = Number(ticker.last);
const change24h = tickerChangePct(ticker);
const volume24h = Number(ticker.volCurrency24h);
const low24 = Number(ticker.low24h);
const high24 = Number(ticker.high24h);

const support = Math.min(ta15.keyLevels.support, ta1h.keyLevels.support, low24);
const resistance = Math.max(ta15.keyLevels.resistance, ta1h.keyLevels.resistance);
const pivot = (high24 + low24 + price) / 3;

let trend = "neutral";
const alignedBull = ta1m.suggestedBias === "long" && ta15.suggestedBias === "long" && ta1h.suggestedBias === "long";
const alignedBear = ta1m.suggestedBias === "short" && ta15.suggestedBias === "short" && ta1h.suggestedBias === "short";
if (alignedBull) trend = "bullish";
else if (alignedBear) trend = "bearish";
else if (ta1h.suggestedBias === "long" && ta1m.suggestedBias === "long") trend = "bullish_pullback";
else if (ta1h.suggestedBias === "short") trend = "bearish";
else trend = "mixed";

const momentumScore = Math.round(((ta1m.momentumScore + ta15.momentumScore + ta1h.momentumScore) / 3) * 1000) / 1000;

let confidence = ta1m.technicalScore * 0.35 + ta15.technicalScore * 0.35 + ta1h.technicalScore * 0.3;
if (spreadPct > 0.5) confidence -= 0.25;
if (fr > 0.005) confidence -= 0.2;
if (fr < -0.005) confidence -= 0.2;
if (alignedBull || alignedBear) confidence += 0.08;
else confidence -= 0.1;
if (change24h > 10) confidence += 0.05;
if (ta1m.rsi14 > 70) confidence -= 0.05;
confidence = Math.max(0, Math.min(1, Math.round(confidence * 1000) / 1000));

let suggestedSide = confidence >= 0.5 ? (ta1m.technicalScore >= ta15.technicalScore ? ta1m.suggestedBias : ta15.suggestedBias) : "neutral";
if (suggestedSide === "neutral" && confidence >= 0.5) suggestedSide = change24h > 0 ? "long" : "short";

const technicalSummary = [
  `1m RSI ${ta1m.rsi14} trend ${ta1m.suggestedBias}; 15m RSI ${ta15.rsi14} trend ${ta15.suggestedBias}; 1h RSI ${ta1h.rsi14} trend ${ta1h.suggestedBias}.`,
  `24h +${change24h.toFixed(2)}% from ${Number(ticker.open24h)}; pullback from high ${high24} to ${price}.`,
  `Funding ${fr.toExponential(2)} (${fb.bias}); spread ${spreadPct.toFixed(3)}%; book imbalance ${ob.imbalance}.`,
  alignedBull || alignedBear ? "Multi-timeframe aligned." : "Timeframes mixed — partial pullback within 24h uptrend.",
].join(" ");

console.log(JSON.stringify({
  instId: "SAGA-USDT",
  price,
  change24h: Math.round(change24h * 1000) / 1000,
  volume24h,
  trend,
  momentumScore,
  keyLevels: [Math.round(support * 1e5) / 1e5, Math.round(pivot * 1e5) / 1e5, Math.round(resistance * 1e5) / 1e5],
  fundingRate: fr,
  fundingBias: fb.bias,
  spreadPct: Math.round(spreadPct * 10000) / 10000,
  technicalSummary,
  suggestedSide: confidence >= 0.5 ? suggestedSide : "neutral",
  confidence,
}));

File: tmp_saga_analyze.py
import json, subprocess

def fetch(method, args):
    p = subprocess.run([
        r"C:\Users\mknig\AppData\Local\Programs\Python\Python312\python.exe",
        r"C:\Users\mknig\owl-swarm\blofin_bridge.py", method, json.dumps(args)
    ], capture_output=True, text=True, cwd=r"C:\Users\mknig\owl-swarm")
    return json.loads(p.stdout.strip())

def rsi(closes, period=14):
    if len(closes) < period + 1:
        return None
    gains, losses = [], []
    for i in range(1, len(closes)):
        d = closes[i] - closes[i-1]
        gains.append(max(d,0)); losses.append(max(-d,0))
    avg_gain = sum(gains[:period]) / period
    avg_loss = sum(losses[:period]) / period
    for i in range(period, len(gains)):
        avg_gain = (avg_gain*(period-1)+gains[i])/period
        avg_loss = (avg_loss*(period-1)+losses[i])/period
    if avg_loss == 0: return 100
    rs = avg_gain/avg_loss
    return 100 - 100/(1+rs)

def trend_score(closes):
    if len(closes) < 5: return 0
    first = sum(closes[:5])/5
    last = sum(closes[-5:])/5
    pct = (last-first)/first*100
    if pct > 0.3: return 1
    if pct < -0.3: return -1
    return 0

def key_levels(candles):
    highs = [float(c[2]) for c in candles]
    lows = [float(c[3]) for c in candles]
    closes = [float(c[4]) for c in candles]
    return {
        'support': min(lows),
        'resistance': max(highs),
        'pivot': (max(highs)+min(lows)+closes[0])/3
    }

for bar, limit in [('1m',50),('15m',30),('1H',24)]:
    c = fetch('get_candles', {'inst_id':'SAGA-USDT','bar':bar,'limit':str(limit)})
    closes = [float(x[4]) for x in reversed(c)]
    print(bar, 'last', closes[-1], 'rsi14', round(rsi(closes),2), 'trend', trend_score(closes), 'chg5', round((closes[-1]-closes[-6])/closes[-6]*100,3) if len(closes)>5 else None)

book = fetch('get_order_book', {'inst_id':'SAGA-USDT'})
bid, ask = float(book['bids'][0][0]), float(ask:=float(book['asks'][0][0]))
mid = (bid+ask)/2
print('spread_pct', round((ask-bid)/mid*100,4))

fund = fetch('get_funding_rate', {'inst_id':'SAGA-USDT'})
print('funding', fund['fundingRate'])

c1h = fetch('get_candles', {'inst_id':'SAGA-USDT','bar':'1H','limit':'24'})
closes1h = [float(x[4]) for x in reversed(c1h)]
print('1h range', min(float(x[3]) for x in c1h), max(float(x[2]) for x in c1h))
print('1h 6h chg', round((closes1h[-1]-closes1h[-7])/closes1h[-7]*100,3))
print('1h 24h chg', round((closes1h[-1]-closes1h[0])/closes1h[0]*100,3))
